<?xml version="1.0" encoding="UTF-8"?>
<launch>
  <arg name="target" default="local" />
  <arg name="move" default="true" />
  <arg name="joy" default="false" />
  <arg name="web" default="false" />
  
  <param name="robot_description" type="str" textfile="$(find srs005_ros)/urdf/robot.urdf"/>
  <arg name="rvizconfig" default="$(find srs005_ros)/rviz/robot.rviz" />

  <arg name="use_sim_time" default="false"/>
  <param name="/use_sim_time" value="$(arg use_sim_time)"/>

  <machine name="local"  address="localhost" />
  <machine name="remote3" address="remote3" env-loader="/home/pi/catkin_ws/devel/env.sh"  user="pi"  password="raspberry"/>
  <machine name="OptiPlex-9010" address="OptiPlex-9010" env-loader="/home/erio/catkin_ws/devel/env.sh" user="erio" password="419000"/>

  <node name="s5_twist_converter" pkg="srs005_ros" type="s5_twist_converter" output="screen">
    <remap from="twist_in"  to="cmd_vel_stamped" />
    <remap from="twist_out" to="cmd_vel" />
  </node>

  <node name="s5_omni_commander" pkg="srs005_ros" type="s5_omni_commander" output="screen">
    <param name="wheel_base" value="0.0972"/>
    <param name="wheel_radius" value="0.019"/>
    <remap from="wheel0" to="wheel0/command" />
    <remap from="wheel1" to="wheel1/command" />
    <remap from="wheel2" to="wheel2/command" />
  </node>

  <node name="s5_omni_odometry" pkg="srs005_ros" type="s5_omni_odometry" output="screen">
    <param name="wheel_base" value="0.0972"/>
    <param name="wheel_radius" value="0.019"/>
    <remap from="wheel0" to="wheel0/position" />
    <remap from="wheel1" to="wheel1/position" />
    <remap from="wheel2" to="wheel2/position" />
  </node>

  <node machine="$(arg target)" name="arduino1" pkg="rosserial_python" type="serial_node.py" args="/dev/ttyACM0" output="screen"/>
  <node machine="$(arg target)" name="uvc_camera_node" pkg="uvc_camera" type="uvc_camera_node" />

  <node machine="$(arg target)" name="ydlidar_node"  pkg="ydlidar"  type="ydlidar_node" output="screen">
    <param name="port"         type="string" value="/dev/ydlidar"/>
    <param name="baudrate"     type="int"    value="115200"/>
    <param name="frame_id"     type="string" value="laser_frame"/>
    <param name="angle_fixed"  type="bool"   value="true"/>
    <param name="low_exposure"  type="bool"   value="false"/>
    <param name="heartbeat"    type="bool"   value="false"/>
    <param name="resolution_fixed"    type="bool"   value="true"/>
    <param name="angle_min"    type="double" value="-180" />
    <param name="angle_max"    type="double" value="180" />
    <param name="range_min"    type="double" value="0.08" />
    <param name="range_max"    type="double" value="16.0" />
    <param name="ignore_array" type="string" value="" />
    <param name="samp_rate"    type="int"    value="9"/>
    <param name="frequency"    type="double" value="7"/>
  </node>

  <node name="s5_odom_tf" pkg="srs005_ros" type="s5_odom_tf" />

  <node name="s5_scan_restamp" pkg="srs005_ros" type="s5_scan_restamp" />

  <node pkg="gmapping" type="slam_gmapping" name="slam_gmapping" clear_params="true">
    <remap from="scan" to="/scan_restamp" />
    <param name="base_frame" value="base_link" />
    <param name="odom_frame" value="odom" />
    <param name="map_update_interval" value="1.0"/>
    <param name="temporalUpdate" value="1.0"/>
    <param name="xmin" value="-5.0"/>
    <param name="ymin" value="-5.0"/>
    <param name="xmax" value="5.0"/>
    <param name="ymax" value="5.0"/>
  </node>

  <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" />
  <node name="rviz" pkg="rviz" type="rviz" args="-d $(arg rvizconfig)" required="true" />

  <group if="$(arg move)">
    <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
      <rosparam file="$(find srs005_ros)/config/costmap_common_params.yaml" command="load" ns="global_costmap" />
      <rosparam file="$(find srs005_ros)/config/costmap_common_params.yaml" command="load" ns="local_costmap" />
      <rosparam file="$(find srs005_ros)/config/local_costmap_params.yaml" command="load" />
      <rosparam file="$(find srs005_ros)/config/global_costmap_params.yaml" command="load" />
      <rosparam file="$(find srs005_ros)/config/base_local_planner_params.yaml" command="load" />
    </node>
  </group>

  <group if="$(arg joy)">
    <node name="joy_node"   pkg="joy" type="joy_node" />
    <node name="s5_twist_publisher" pkg="srs005_ros" type="s5_twist_publisher">
      <param name="x_scale" value="0.3" />
      <param name="y_scale" value="0.0" />
      <param name="yaw_scale" value="1.0" />
    </node>
  </group>

  <group if="$(arg web)">
    <include file="$(find roswww)/launch/roswww.launch" />
    <node pkg="web_video_server" type="web_video_server" name="web_video_server" />
    <include file="$(find rosbridge_server)/launch/rosbridge_websocket.launch" />
    <node name="s5_twist_publisher" pkg="srs005_ros" type="s5_twist_publisher">
      <param name="x_scale" value="0.3" />
      <param name="y_scale" value="0.0" />
      <param name="yaw_scale" value="-1.0" />
    </node>
  </group>
</launch>
